Userland/Applications/PixelFlutClient/Client.cpp

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright (c) 2023, kleines Filmröllchen <filmroellchen@serenityos.org>.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "Client.h"
+#include <LibImageDecoderClient/Client.h>
+
+constexpr StringView get_command = "SIZE\n"sv;
+
+ErrorOr<NonnullRefPtr<Client>> Client::create(StringView image_path, StringView server)
+{
+    // Extract hostname and port and connect to server.
+    auto parts = server.split_view(':');
+    auto maybe_port = parts.take_last().to_uint();
+    if (!maybe_port.has_value())
+        return Error::from_string_view("Invalid port number"sv);
+
+    auto port = maybe_port.release_value();
+    auto hostname = TRY(String::join(':', parts));
+
+    auto socket = TRY(Core::BufferedTCPSocket::create(TRY(Core::TCPSocket::connect(hostname.to_deprecated_string(), port))));
+
+    // Ask the server for the canvas size.
+    TRY(socket->write_until_depleted(get_command.bytes()));
+    auto buffer = TRY(ByteBuffer::create_zeroed(KiB));
+    auto size_line = TRY(socket->read_line(buffer));
+    if (!size_line.starts_with("SIZE"sv))
+        return Error::from_string_view("Server didn't return size correctly"sv);
+
+    auto size_parts = size_line.split_view(' ');
+    auto maybe_width = size_parts[1].to_uint();
+    auto maybe_height = size_parts[2].to_uint();
+    if (!maybe_width.has_value() || !maybe_height.has_value())
+        return Error::from_string_view("Width or height invalid"sv);
+
+    auto width = maybe_width.release_value();
+    auto height = maybe_height.release_value();
+    Gfx::IntSize canvas_size = { width, height };
+
+    // Read input image.
+    auto image_file = TRY(Core::File::open(image_path, Core::File::OpenMode::Read));
+    auto image_decoder = TRY(ImageDecoderClient::Client::try_create());
+    ScopeGuard guard = [&] { image_decoder->shutdown(); };
+
+    auto byte_buffer = TRY(image_file->read_until_eof(16 * KiB));
+    auto maybe_image = image_decoder->decode_image(byte_buffer);
+
+    if (!maybe_image.has_value())
+        return Error::from_string_view("Image could not be read"sv);
+
+    auto image = maybe_image->frames.take_first().bitmap.release_nonnull();
+
+    // Make sure to not draw out of bounds; some servers will disconnect us for that!
+    if (image->width() > canvas_size.width()) {
+        auto fitting_scale = static_cast<float>(canvas_size.width()) / image->width();
+        image = TRY(image->scaled(fitting_scale, fitting_scale));
+    }
+
+    return TRY(adopt_nonnull_ref_or_enomem(new (nothrow) Client(move(socket), move(image), canvas_size)));
+}
+
+Client::Client(NonnullOwnPtr<Core::BufferedTCPSocket> socket, NonnullRefPtr<Gfx::Bitmap> image, Gfx::IntSize canvas_size)
+    : m_socket(move(socket))
+    , m_image(move(image))
+    , m_canvas_size(canvas_size)
+{
+    dbgln("Connected to server, image {}, canvas size {}", m_image, m_canvas_size);
```

Comment:
Maybe just make this an `outln` so you can see things happening in the terminal?